### ISSUE TITLE
Update `use_build_context_synchronously` for `BuildContext.mounted`

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/linter#3688]
In https://github.com/flutter/flutter/pull/111619 we are introducing a new `BuildContext.mounted` property. This makes it easier to use `BuildContext` error-free across asynchronous gaps in `StatelessWidgets` and other instances where you can't check `State.mounted`. This change will require an update to the `use_build_context_synchronously` lint. The following usage of the new property should not trigger any lints:

```dart
  Widget build(BuildContext context) {
    return OutlinedButton(
      onPressed: () async {
        await Future.delayed(const Duration(seconds: 1));
        if (context.mounted) {
          Navigator.of(context).pop(); // No lint here, the usage after the async gap is properly guarded by `context.mounted`.
        }
      },
      child: const Text('Delayed pop'),
    );
  }
```

Furthermore, the documentation of the lint should be updated to make `context.mounted` the recommended way of fixing the lint warning. (`State.mounted` should probably just be eliminated from the docs)

**Comment by goderbauer** [comment]
cc @a14n @pq Any chance one of you would be interested in doing this? 😄 

I wonder what the best sequence of events would be? Should we wait for the lint fix to roll into flutter before submitting the flutter PR?

**Comment by a14n** [comment]
If I read the code correctly I think it should already be supported and no change is really needed.

**Comment by goderbauer** [comment]
> 
If I read the code correctly I think it should already be supported and no change is really needed.

You're right. I just checked it with my implementation in https://github.com/flutter/flutter/pull/111619 and the lint doesn't actually trigger. So, we just need to update the docs. Thanks for putting up a PR for that.

**Comment by pq** [comment]
> If I read the code correctly I think it should already be supported and no change is really needed.

💯 

Thanks for the doc update! 👍

**Comment by Almighty-Alpaca** [comment]
I just tried out the latest Flutter master version (Flutter 3.6.0-4.0.pre.11, Analyzer 
From my limited testing this only works half of the time currently.

Changing the example to 
```dart
  Widget build(BuildContext context) {
    return OutlinedButton(
      onPressed: () async {
        await Future<void>.delayed(const Duration(seconds: 1));
        if (!context.mounted) {
          return;
        }

        Navigator.of(context).pop();
      },
      child: const Text('Delayed pop'),
    );
  }
```
produces one `use_build_context_synchronously` lint violation for `context.mounted`, while the following
```dart
  Widget build(BuildContext context) {
    return OutlinedButton(
      onPressed: () async {
        await Future.delayed(const Duration(seconds: 1));
        if (!context.mounted) {
          // Do something else
        } else {
          Navigator.of(context).pop();
        }
      },
      child: const Text('Delayed pop'),
    );
  }
  ```
Produces two lint violations, one for `context.mounted` and one for `Navigator.of(context).pop()`.
